### DISCUSSION TITLE
Tools > Key Generator ?

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#301] [Q&A]
Hi everybody,
I just wanted to know if someone could explain me what's  the Key Generator (Tools) exactly for, and how it works in practice?

Thanks
 
Damien

**Comment by tomlem** [comment]
RSA key is a private key based on RSA algorithm. Private Key is used for authentication and a symmetric key exchange during establishment of an SSL/TLS session. 

Public key authentication is a way of logging into an SSH/SFTP account using a cryptographic key rather than a password.

If you use very strong SSH/SFTP passwords, your accounts are already safe from brute force attacks. However, using public key authentication provides many benefits when working with multiple developers. For example, with SSH keys you can

1.     allow multiple developers to log in as the same system user without having to share a single password between them;
2.     revoke a single developer's access without revoking access by other developers; and
3.     make it easier for a single developer to log in to many accounts without needing to manage many different passwords.

https://www.namecheap.com/support/knowledgebase/article.aspx/798/67/what-is-an-rsa-key-used-for/

https://www.cryptosys.net/pki/rsakeyformats.html

https://serverpilot.io/docs/how-to-use-ssh-public-key-authentication/

**Reply by damiensellier** [reply]
Thanks tomlem. I know what RSA is but the Key Generation via Juce class generateKeyFile() is not that obvious. I don't even know if it's fully suported by Ctrlr. Hence my initial question.

**Reply by Godlike-Productions** [reply]
I think it's the basis of  copy protection, but could also be used to establish secure web sessions for product updates.  I haven't gone near that rabbit hole yet.  Still haven't finished falling down the ones that I'm in.

**Reply by damiensellier** [reply]
Thanks @Godlike-Productions  I think you are right. I'm currently reading the JUCE documentation about copy protection and I think this "Key Generator" feature in ctrlr is relying on JUCE KeyGeneration Class.
[JUCE KeyGeneration Class Reference](https://docs.juce.com/master/classKeyGeneration.html)

From your Jupiter8 Morph page on your website you say : "The plugin is protected by PACE iLok technology, though the plugin can be used without a physical iLok."  I can imagine you found a way to make it work :)
Is the whole app or VST embedded in a secured wrapper? or is it a function called internally that do the licence check?
Pace Website is not documented, there's even no way to check their prices for their licence management service.
I really don't know how to work with that and even where to start. If you could help with that, that would be really helpful. 
Thanks ;)
Damien

**Reply by Godlike-Productions** [reply]
PACE provides tools to wrap the binary, so as long as you have binaries that run, you run PACE's copy protection tool and they add their security layer around the outside.  They have a more secure option for securing the code internally, but I'm really interested in just stopping people sharing the dll file and actually having to buy a registered copy.  The number of Jupiter 8's out there is pretty damn limited, and the time required to build this from the sysex I've used is worth way more than the $79 I'm charging

Basically I don't need to use the KeyGeneration within Juce.  PACE does that externally to Ctrlr.  Each binary is wrapped seperately, so I wrap the AU, the windows VST, windows exe each as essentially a different product, but I use the same configuration, so a registered key for one of the will allow any of them to run. Each binary has the same lock on it, essentially.

Bear in mind that PACE does charge a fee for every key issued, which needs to be accounted for in product pricing. It's significant, but fair IMHO, but I can't divulge the figure because of the contract.  Also, it means that I can leave securing of the product to the experts and use a pretty standardized approach with the minimum of hassle for me.  It's quite easy to automate the whole sales and registration process, if you choose the right online store provider.

I'm happy to work through with anyone who has a contract with PACE (or a trial) on how to secure a ctrlr instance with iLok, though iLok's support personnel have been pretty good.

**Reply by damiensellier** [reply]
@Godlike-Productions I really appreciate your insight about that. There's nothing on the internet about that probably because of NDA. I'll get in touch with PACE in a next future.
I finally found their page to explain their process : [Pace Pro Audio](https://www.paceap.com/proaudio.html)

The included RSA protection within JUCE is pretty cool too and can probably deter some hackers. There's a unique user ID file generated on the client side and a JUCE class that calls a php server to request authorization. It's working exactly like the one from plugin-alliance. You can authorize either offline with your unique file or online from a server. 

There's a JUCE tutorial about that : 

[Tutorial: Unlock your plugins through online registration ](https://docs.juce.com/master/tutorial_online_unlock_status.html)
I need to check if Ctrlr has all the required class and try.

Thanks @Godlike-Productions  It's very appreciated.

**Reply by Godlike-Productions** [reply]
I think that using the Juce method you will need to manage your own server. While it's something I could probably accomplish, I wouldn't be able to do it in the timeframe I got iLok working in.  I will look into it in the future though.  Thanks for the link.
